crates/bevy_ecs/src/bundle/mod.rs

```diff
@@ -244,7 +244,13 @@ pub trait DynamicBundle {
     /// Calls `func` on each value, in the order of this bundle's [`Component`]s. This passes
     /// ownership of the component values to `func`.
     #[doc(hidden)]
-    fn get_components(self, func: &mut impl FnMut(StorageType, OwningPtr<'_>)) -> Self::Effect;
+    unsafe fn get_components(ptr: *const Self, func: &mut impl FnMut(StorageType, OwningPtr<'_>));
+
+    // SAFETY:
+    // - Must be called exactly once after `get_components` has been called.
+    // - `ptr` must point to a valid instance of `Self` but does not necessary need to be aligned.
+    #[doc(hidden)]
+    unsafe fn apply_effect(ptr: *const Self, entity: &mut EntityWorldMut);
```

Comment:
I'm not sure how `ptr` can point to a valid instance of `Self` when calling `apply_effect`. Supposedly most of `Self` will be logicaly moved out in `get_components`, so `ptr` won't point to a valid instance of `Self` here. Rather only the bits that weren't moved out in `get_components` can reasonably still be valid.